sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ColumnResolutionHelper.scala

```diff
@@ -487,42 +487,76 @@ trait ColumnResolutionHelper extends Logging with DataTypeErrorsBase {
   //       original expression as it is.
   private def tryResolveColumnByPlanId(
       e: Expression,
-      q: LogicalPlan,
-      idToPlan: mutable.HashMap[Long, LogicalPlan] = mutable.HashMap.empty): Expression = e match {
+      q: LogicalPlan): Expression = e match {
     case u: UnresolvedAttribute =>
-      resolveUnresolvedAttributeByPlanId(
-        u, q, idToPlan: mutable.HashMap[Long, LogicalPlan]
-      ).getOrElse(u)
+      resolveUnresolvedAttributeByPlanId(u, q).getOrElse(u)
     case _ if e.containsPattern(UNRESOLVED_ATTRIBUTE) =>
-      e.mapChildren(c => tryResolveColumnByPlanId(c, q, idToPlan))
+      e.mapChildren(c => tryResolveColumnByPlanId(c, q))
     case _ => e
   }
 
   private def resolveUnresolvedAttributeByPlanId(
       u: UnresolvedAttribute,
-      q: LogicalPlan,
-      idToPlan: mutable.HashMap[Long, LogicalPlan]): Option[NamedExpression] = {
+      q: LogicalPlan): Option[NamedExpression] = {
     val planIdOpt = u.getTagValue(LogicalPlan.PLAN_ID_TAG)
     if (planIdOpt.isEmpty) return None
     val planId = planIdOpt.get
     logDebug(s"Extract plan_id $planId from $u")
 
-    val plan = idToPlan.getOrElseUpdate(planId, {
-      findPlanById(u, planId, q).getOrElse {
-        // For example:
-        //  df1 = spark.createDataFrame([Row(a = 1, b = 2, c = 3)]])
-        //  df2 = spark.createDataFrame([Row(a = 1, b = 2)]])
-        //  df1.select(df2.a)   <-   illegal reference df2.a
+    val isMetadataAccess = u.getTagValue(LogicalPlan.IS_METADATA_COL).isDefined
+
+    val resolved = resolveUnresolvedAttributeByPlanId(u, planId, isMetadataAccess, q)
+    if (resolved.isEmpty) {
+      // For example:
+      //  df1 = spark.createDataFrame([Row(a = 1, b = 2, c = 3)]])
+      //  df2 = spark.createDataFrame([Row(a = 1, b = 2)]])
+      //  df1.select(df2.a)   <-   illegal reference df2.a
+      throw new AnalysisException(
+        errorClass = "CANNOT_RESOLVE_WITH_PLAN_ID",
+        messageParameters = Map(
+          "expression" -> toSQLId(u.nameParts),
+          "id" -> planId.toString,
+          "plan" -> q.toString
+        ),
+        origin = u.origin
+      )
+    }
+    resolved
+  }
+
+  private def resolveUnresolvedAttributeByPlanId(
+      u: UnresolvedAttribute,
+      id: Long,
+      isMetadataAccess: Boolean,
+      p: LogicalPlan): Option[NamedExpression] = {
+    if (p.getTagValue(LogicalPlan.PLAN_ID_TAG).contains(id)) {
+        resolveUnresolvedAttributeByPlan(u, p, isMetadataAccess)
+    } else {
+      val candidates = p.children.flatMap { child =>
+        val outputSet = if (isMetadataAccess) {
+          // NOTE: A metadata column might appear in `output` instead of `metadataOutput`.
+          child.outputSet ++ AttributeSet(child.metadataOutput)
+        } else {
+          child.outputSet
+        }
+        resolveUnresolvedAttributeByPlanId(u, id, isMetadataAccess, child)
```

Comment:
We can assume that `resolveUnresolvedAttributeByPlanId` always return valid column, then the code can be simplified
```
val candidates = p.children.flatMap(resolveUnresolvedAttributeByPlanId...)
if (candidates.length > 1) fail ...
val outputSet = if (isMetadataAccess) p.outputSet ++ ... else p.outputSet
candidates.filter(_.references.subsetOf(outputSet))
```

Reply:
I guess here should use child's outputSet instead of `p`

Reply:
I think it could be
```
val candidates = p.children.flatMap(resolveUnresolvedAttributeByPlanId...)
if (candidates.isEmpty) {
  None
} else {
  val outputSet = ...
  candidates.filter(_.references.subsetOf(outputSet))
}
```

to only new a `outputSet` for non-empty candidates